File: learn-data_structures/python/linked-list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    # Add to the end
    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    # Remove from start
    def remove_first(self):
        if not self.head:
            return
        self.head = self.head.next

    # Display list
    def print_list(self):
        current = self.head
        while current:
            print(current.data)
            current = current.next


# Usage
my_list = LinkedList()
my_list.append(1)
my_list.append(2)
my_list.append(3)
my_list.remove_first()
my_list.print_list()

File: learn-algos/src/linked-list.ts
class ListNode<T> {
  constructor(public data: T, public next: ListNode<T> | null = null) {}
}

class LinkedList<T> {
  private head: ListNode<T> | null = null;

  // Add to the end
  append(data: T): void {
    const newNode = new ListNode(data);
    if (!this.head) {
      this.head = newNode;
      return;
    }
    let current = this.head;
    while (current.next) {
      current = current.next;
    }
    current.next = newNode;
  }

  // Remove from start
  removeFirst(): void {
    if (!this.head) return;
    this.head = this.head.next;
  }

  // Display list
  printList(): void {
    let current = this.head;
    while (current) {
      console.log(current.data);
      current = current.next;
    }
  }
}

// Usage
const myList = new LinkedList<number>();
myList.append(1);
myList.append(2);
myList.append(3);
myList.removeFirst();
myList.printList();
